### PR TITLE
#128 migrate project to pyrpoject toml based package setup

### LINKED ISSUE #128: Migrate project to a pyproject.toml based package setup
## Why
Currently the project packaging, configuration and dependencies are managed using a mix of input sources.
E.g. dependencies are defined in `setup.py` and in various requirements files (`requirements.txt` etc.).
Modern python projects use a common "project-file" which takes care about lots of things which have been
done by multiple inputs like (`MANIFEST.in, setup.cfg, setup.py, requirements.txt, ...`).

Switching to `pyproject.toml` therefore will simplify maintenance and release efforts of the package.

For further details see also: 
* [Poetry](https://python-poetry.org/)
* [PEP 517](https://peps.python.org/pep-0517/)
* [PEP 518](https://peps.python.org/pep-0518/)
* [PEP 621](https://peps.python.org/pep-0621/)
* [PEP 660](https://peps.python.org/pep-0660/)

## Tasks
- [x] Add `pyproject.toml`
  - [x] Add requirements 
  - [x] Add dev requirements
  - [x] Add extra requirements
  - [x] Add test requirements
- [x] Add poetry.lock
- [x] Move configuration(s) into `pyproject.toml`
  - [x] setup.cfg
  - [x] MANIFEST.in
- [x] Remove .ptup.yaml
- [x] Add support for Dependabot
- [x] Change GH-Action to build and deploy using poetry

### LINKED ISSUE #135: Remove support for Python < 3.8
This project supports python version >= 3.7. We want to remove older python versions. 
* Remove support for python version < 3.8

### LINKED ISSUE #146: turbodbc feature can't be used for poetry based project setup
## Description
Due to differences in the structure/naming of build folder(s) and artifacts when using `poetry` instead of `pip`, building the `turbodbc` src package will fail. For further details see https://github.com/blue-yonder/turbodbc/issues/358.

Additional complexity to this problem may be added by the fact that for the `turbodbc` project it may not be possible or feasible to back port the fix and therefore a version update of the `turbodbc` dependency may also be needed.

`requirements_extras.txt`
> turbodbc==3.3.0 # 4.0.0 introduced some breaking change (e.g. drops Python 2.7 support) if the fix is not backported

## Possible Solution(s)
* Drop turbodbc feature/support
* The turbodbc project also provides a wheel for the different target systems (additional to the src distribution)
* The turbodbc project fixes the deduction of the build- paths/artifacts
   * Update the version of the dependency if the fix isn't back ported
     (Due to the fact that the latest turbodbc package requires `python >= 3.8` this implies we also address exasol/sqlalchemy-exasol#135)

## Tasks
- [x] re-enable/add turbodbc support once https://github.com/blue-yonder/turbodbc/issues/358 is fixed

### PR DESCRIPTION
**Pull request by Nicoretti**
fixes #128

### CODE REVIEW

File: scripts/version_check.py [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+
+Version = namedtuple("Version", ["major", "minor", "patch"])
+
+_SUCCESS = 0
+_FAILURE = 1
+
+# fmt: off
+_VERSION_MODULE_TEMPLATE = cleandoc('''
+# ATTENTION: 
+# This file is generated, do not edit this file manually!
```
Comment (kaklakariada) [inline scripts/version_check.py]:
```suggestion
# This file is generated, do not edit it manually!
```

File: scripts/version_check.py [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+# If you need to change the version, do so in the project.toml, e.g. by using `poetry version X.Y.Z`.
```
Comment (kaklakariada) [inline scripts/version_check.py:18]: 👍

File: sqlalchemy_exasol/version.py [hunk 3]
```diff
@@ -0,0 +1,9 @@
+# ATTENTION: 
+# This file is generated, do not edit this file manually!
+# If you need to change the version, do so in the project.toml, e.g. by using `poetry version X.Y.Z`.
```
Comment (kaklakariada) [inline sqlalchemy_exasol/version.py:3]: Do we need to check in this generated file or is it possible to create it at build time? Or does the pre-commit-hook do this automatically?

**Review by kaklakariada — CHANGES_REQUESTED** [review]

File: sqlalchemy_exasol/version.py (hunk 3 above)
Comment (Nicoretti) [reply sqlalchemy_exasol/version.py:3]: Very good point! I have thought about this quite a bit also because the context is quite complex.
The TL;DR more or less is this:
* It is not as straight forward to do it like that as within the old "build system"
* It can be done with package metadata though see [this comment](https://github.com/python-poetry/poetry/pull/2366#issuecomment-652418094)
* Using metadata prior to `python 3.8` has quite some pitfalls
  ("Dependency hell of the metadata-lib backport")
  
So my plan was: for now add a git hook and verification step to the build and when the library is [upgraded to 3.8](https://github.com/exasol/sqlalchemy-exasol/issues/135#issuecomment-1143186610) switch to the metadatalib variant.

Comment (tkilias) [reply sqlalchemy_exasol/version.py:3]: Regarding the upgrade to python 3.8, I think this project can wait longer with that, we only need to drop 3.6 if didn't do it yet. And, please don't mix things in issues together. The version file is a separate issue. I guess you want to use importlib.metadata, for pre 3.8 there is a Backport package, you can use. We already used it already various other projects.

Comment (tkilias) [reply sqlalchemy_exasol/version.py:3]: However, there is another issue with importlin_metadata. From what we saw, it only works for installed packages.

File: setup.cfg [hunk 4]
```diff
@@ -1,6 +1,3 @@
-[egg_info]
-#tag_build = dev
-
 [tool:pytest]
```
Comment (tkilias) [inline setup.cfg]: I think, this can move as well https://docs.pytest.org/en/6.2.x/customize.html#configuration-file-formats

**Review by tkilias — CHANGES_REQUESTED** [review]

File: setup.cfg (hunk 4 above)
Comment (Nicoretti) [reply setup.cfg]: @tkilias sadly no, I tried (for details see https://github.com/exasol/sqlalchemy-exasol/issues/144)

Comment (tkilias) [reply setup.cfg]: Ok, good to know, than I suggest add a comment

File: sqlalchemy_exasol/version.py (hunk 3 above)
Comment (Nicoretti) [reply sqlalchemy_exasol/version.py:3]: @tkilias that it works only on installed package is fine in the scenario of the sqlaclhemy plugin I would say (I don't expect the extension points to work without installation, I am not 100% sure though).

Regarding `importlib.metadata` previous to 3.8 where it is part of the standard library can be hell of a mess if more than one library uses it, which is often the case (We don't want to run into dependency issues because of the version imho).

File: scripts/version_check.py (hunk 1 above)
Comment (Nicoretti) [reply scripts/version_check.py]: Done

File: setup.cfg (hunk 4 above)
Comment (Nicoretti) [reply setup.cfg]: Good point, I will do so.

File: .github/workflows/CI.yml [hunk 5]
```diff
@@ -35,26 +35,24 @@ jobs:
     - name: Setup integration-test-docker-environment
       uses: actions/setup-python@v2
 
+    - name: Install via apt
+      run: sudo apt-get install unixodbc unixodbc-dev libboost-date-time-dev libboost-locale-dev libboost-system-dev
+
     - name: Setup Python ${{ matrix.python }}
       uses: actions/setup-python@v2
       with:
         python-version: ${{ matrix.python }}
```
Comment (kaklakariada) [inline .github/workflows/CI.yml]:
```suggestion
    - name: Setup Python ${{ matrix.python }}
      uses: actions/setup-python@v3
      with:
        python-version: ${{ matrix.python }}
        cache: 'poetry'
```

see https://github.com/actions/setup-python#caching-packages-dependencies

File: .github/workflows/CI.yml [hunk 6]
```diff
@@ -35,26 +35,24 @@ jobs:
     - name: Setup integration-test-docker-environment
       uses: actions/setup-python@v2
```
Comment (kaklakariada) [inline .github/workflows/CI.yml]:
```suggestion
```

is the setup-python action duplicated?

File: .github/workflows/CI.yml [hunk 7]
```diff
@@ -94,11 +90,10 @@ jobs:
       with:
         python-version: ${{ matrix.python }}
```
Comment (kaklakariada) [inline .github/workflows/CI.yml:93]: Also use version v3 and activate caching for poetry.

**Review by kaklakariada — CHANGES_REQUESTED** [review]

File: .github/workflows/CI.yml [hunk 8]
```diff
@@ -65,21 +63,19 @@ jobs:
         ITDE_TAG: "0.10.0"
 
     - name: Check documentation links ${{ matrix.python }} using ${{ matrix.connector }}
-      run: nox -s "check-links"
+      run: poetry run nox -s "check-links"
```
Comment (tkilias) [inline .github/workflows/CI.yml]: you might want to move this to its own workflow, because check links can be flaky and you don't want to prevent the test running because some links don't respond properly

File: .pre-commit-config.yaml [hunk 9]
```diff
@@ -0,0 +1,11 @@
...
+repos:
+- repo: local
+  hooks:
+
+    - id: version-check
+      name: Check if version(s) are in sync
+      always_run: true
+      verbose: true
+      language: system
+      entry: python scripts/version_check.py sqlalchemy_exasol/version.py
```
Comment (tkilias) [inline .pre-commit-config.yaml]: Theoretically, you could simply use the nox command check_version

File: sqlalchemy_exasol/pyodbc.py [hunk 10]
```diff
@@ -9,7 +9,7 @@
 import re
 import sys
 import logging
-from distutils.version import LooseVersion
+from packaging import version
```
Comment (tkilias) [inline sqlalchemy_exasol/pyodbc.py:12]: is this an additional package, if yes then add it as explicit dependency, because I think, it is not listed, yet. However, I like that we replace distutils

**Review by tkilias — CHANGES_REQUESTED** [review]

File: sqlalchemy_exasol/pyodbc.py (hunk 10 above)
Comment (Nicoretti) [reply sqlalchemy_exasol/pyodbc.py:12]: Good catch, thx! Did not list it explicitly!

File: .pre-commit-config.yaml (hunk 9 above)
Comment (Nicoretti) [reply .pre-commit-config.yaml]: Yes, but for now I wanted nox to check and for the hook I wanted it to fix it if necessary.

File: .github/workflows/ci.yaml [hunk 11]
```diff
@@ -65,21 +63,19 @@ jobs:
         ITDE_TAG: "0.10.0"
 
     - name: Check documentation links ${{ matrix.python }} using ${{ matrix.connector }}
-      run: nox -s "check-links"
+      run: poetry run nox -s "check-links"
```
Comment (tkilias) [inline .github/workflows/ci.yaml]: You run this already in an extra workflow, you can remove this, such that we don't fail the test, because some links don't work

File: pyproject.toml [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+]
+exclude = []
+
+
+[tool.poetry.dependencies]
+python = ">=3.8,<4.0"
+SQLAlchemy = ">=1.3.24,<1.4"
+pyodbc = ">=4.0.32"
+packaging = "^21.3"
+turbodbc = {path = "../../Downloads/turbodbc-4.5.3.tar.gz"}
```
Comment (tkilias) [inline pyproject.toml]: this needs to be changed later

**Review by tkilias — CHANGES_REQUESTED** [review]

File: pyproject.toml (hunk 12 above)
Comment (Nicoretti) [reply pyproject.toml]: :white_check_mark:  agreed, currently I need it because I wait on the turbodbc fix.
(will be removed/changed in the final version)

File: .github/workflows/ci.yaml (hunk 11 above)
Comment (Nicoretti) [reply .github/workflows/ci.yaml]: good catch thx!

File: noxfile.py [hunk 13]
```diff
@@ -83,20 +86,57 @@ def temporary_odbc_config(config):
...
+        args.append(cli_args.username)
+    if cli_args.password:
+        args.append("--password")
+        args.append(cli_args.password)
+
+    version_file = version_from_python_module(PROJECT_ROOT / 'sqlalchemy_exasol' / 'version.py')
+    module_version = version_from_poetry()
+    git_version = version_from_string(tags()[-1])
+
+    if not (module_version == git_version == version_file):
```
Comment (tkilias) [inline noxfile.py]: Are you sure this is correct?

File: noxfile.py [hunk 14]
```diff
@@ -83,20 +86,57 @@ def temporary_odbc_config(config):
...
+
+    args = []
+    parser = create_parser()
+    cli_args = parser.parse_args(session.posargs)
+    if cli_args.dry_run:
+        args.append("--dry-run")
+    if cli_args.username:
+        args.append("--username")
+        args.append(cli_args.username)
+    if cli_args.password:
```
Comment (tkilias) [inline noxfile.py]: Do we need a interactive password request?

File: CHANGES.md [hunk 15]
```diff
@@ -1,3 +1,11 @@
+# 3.0.0
+* Dropped python 3.7 support
+  * If you still depend on python 3.7 use the 2.x version line
+* Updated dependencies, most noticeably
+  * turbodbc to 4.x.y
```
Comment (tkilias) [inline CHANGES.md]: Hm, is this still valid

File: .github/workflows/ci-cd.yml [hunk 16]
```diff
@@ -91,17 +92,10 @@ jobs:
...
-      uses: pypa/gh-action-pypi-publish@release/v1
+    - name: Install poetry
+      uses: abatilo/actions-poetry@v2.0.0
       with:
-        user: __token__
-        password: ${{ secrets.pypi_token }}
+        poetry-version: 1.1.13
+
+    - name: Build and push package to PYPI
+      run: poetry run nox -s release -- --username=__token__ --password=${{ secrets.pypi_token }}
```
Comment (tkilias) [inline .github/workflows/ci-cd.yml]: Is this best practices to Adresse the secret this way

File: pyproject.toml [hunk 17]
```diff
@@ -0,0 +1,80 @@
...
+    "Operating System :: POSIX :: Linux",
+    "Intended Audience :: Developers",
+    "Intended Audience :: Education",
+    "Intended Audience :: Information Technology",
+    "License :: OSI Approved :: BSD License",
+    "Programming Language :: SQL",
+    "Programming Language :: Python :: 3",
+    "Programming Language :: Python :: 3.7",
+    "Programming Language :: Python :: 3.8",
+    "Programming Language :: Python :: 3.9",
```
Comment (tkilias) [inline pyproject.toml:27]: 3.10 is missing

**Review by tkilias — CHANGES_REQUESTED** [review]

File: noxfile.py (hunk 13 above)
Comment (Nicoretti) [reply noxfile.py]: I think so, should be only triggered on tagging/tag push

File: pyproject.toml (hunk 17 above)
Comment (Nicoretti) [reply pyproject.toml:27]: Good catch!

File: noxfile.py (hunk 14 above)
Comment (Nicoretti) [reply noxfile.py]: Yes would make sense, at least when you run it locally, generally it's not recommended to have the pw within a cli call.

File: odbcconfig/odbcinst.ini [hunk 18]
```diff
@@ -4,4 +4,4 @@
 
 [EXAODBC]
 #Driver location will be appended in build environment:
-#DRIVER=<PATH TO DRIVER>
+DRIVER=/home/nic/Projects/sqlalchemy-exasol/driver/libexaodbc-uo2214lv1.so
```
Comment (tkilias) [inline odbcconfig/odbcinst.ini]: Local path. Should we provide a template file instead and add this file to gitignore

**Review by tkilias — CHANGES_REQUESTED** [review]

File: .github/workflows/ci-cd.yml (hunk 16 above)
Comment (Nicoretti) [reply .github/workflows/ci-cd.yml]: I removed the password flag, I can see that it could be used locally by accident. Switched to env var only set during exactly the publish invocation. Poetry [config](https://python-poetry.org/docs/repositories/#configuring-credentials) supports keyring, but not sure if that's in the image out of the box and if it fails it will fall back to write it to a file.

File: odbcconfig/odbcinst.ini (hunk 18 above)
Comment (Nicoretti) [reply odbcconfig/odbcinst.ini]: Good point, did not happen the first time!

File: .github/workflows/ci-cd.yml (hunk 16 above)
Comment (tkilias) [reply .github/workflows/ci-cd.yml]: That is actually better, because we usually only want to release from the CI

File: .github/workflows/ci-cd.yml [hunk 19]
```diff
@@ -98,4 +98,4 @@ jobs:
         poetry-version: 1.1.13
 
     - name: Build and push package to PYPI
-      run: poetry run nox -s release -- --username=__token__ --password=${{ secrets.pypi_token }}
+      run: POETRY_PYPI_TOKEN_PYPI="${{ secrets.pypi_token }}" poetry run nox -s release
```
Comment (tkilias) [inline .github/workflows/ci-cd.yml]: Use env property of GitHub action

Comment (tkilias) [reply .github/workflows/ci-cd.yml]: Can we try this, I saw some issues in the poetry repos for the env variable

File: odbcconfig/odbcinst.ini.template [hunk 20]
```diff
@@ -0,0 +1,7 @@
+[ODBC]
+#Trace=yes
+#TraceFile=~/odbc.trace
+
+[EXAODBC]
+#Driver location needs to be added in build environment:
+DRIVER=<PATH_TO_DRIVER>
```
Comment (tkilias) [inline odbcconfig/odbcinst.ini.template:7]: This template is used by a user and not a program or? If it would be used by a program, we should use Jinja or so

**Review by tkilias — CHANGES_REQUESTED** [review]

File: odbcconfig/odbcinst.ini.template (hunk 20 above)
Comment (Nicoretti) [reply odbcconfig/odbcinst.ini.template:7]: User

File: .github/workflows/ci-cd.yml (hunk 19 above)
Comment (Nicoretti) [reply .github/workflows/ci-cd.yml]: Can you elaborate what you mean exactly?

**Review by tkilias — DISMISSED** [review]

**Review by tkilias — APPROVED** [review]

**Review by kaklakariada — APPROVED** [review]
